dgram: improve "address" parameter behavior in Socket.prototype.send

boneskull opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)

dgram

##### Description of change

When using `Socket.prototype.send`, I was confused about when `address` was required, since the docs are ambiguous at best, and conflicting at worst: they claim `address` will default to `127.0.0.1`, *but* this is only in the case where `callback` is not supplied.

My confusion manifest in an exception from the `dns` module, which was unhelpful:

```
dns.js:112
    throw new TypeError('Invalid arguments: ' +
    ^

TypeError: Invalid arguments: hostname must be a string or falsey
    at Object.lookup (dns.js:112:11)
    at lookup (dgram.js:27:14)
    at UDP.lookup4 [as lookup] (dgram.js:32:10)
    at Socket.send (dgram.js:362:16)
```

`dns.lookup` expects a valid `hostname` argument, yet this parameter name ("hostname") is not used in `Socket.prototype.send`.  I thought it better to fail with helpful error messaging *before* getting as far as `dns.lookup`, which prompted this PR. 

~~The signature of `Socket.prototype.send` is a bit awkward, and perhaps the *best* fix here is to change it.  I see this PR as being an *improvement* nonetheless.  Description of changes follow.~~

**UPDATE Jan 8 17**

Upon suggestion by @mscdex, I've modified the changes to make `address` optional in all cases. 

Summary of changes:

- In `Socket.prototype.send`, `address` is now *always* optional; it is no longer required to use `callback`.  The signature has changed from this:
  ```
  socket.send(msg, [offset, length,] port, address [, callback])
  ```
  to this:
  ```
  socket.send(msg, [offset, length,] port, [, address] [, callback])
  ```
- Faster failure and more helpful error messaging in case when `address`  argument is present but invalid; exceptions no longer fall through  to `dns` module.
- Updates docs, adds tests.

mscdex commented:
I think it would probably be better to just support the 'callback without address' case instead of adding a note/warning in the documentation.

boneskull commented:
@mscdex I agree.  I don't see any technical reason why we couldn't.  I didn't want to be presumptuous.

I'll update this PR accordingly.

boneskull commented:
@mscdex I've updated this PR as suggested, rebased, etc.  My local tests now pass 100%.

mscdex commented on lib/dgram.js:
```diff
@@ -344,6 +346,13 @@ Socket.prototype.send = function(buffer,
+  if (typeof address === 'function') {
+    callback = address;
+    address = undefined;
+  } else if (typeof address !== 'undefined' && typeof address !== 'string') {
```
We can use `address !== undefined` instead of `typeof address !== 'undefined'`.

mscdex replied:
I don't think it matters in this case since it would just fail the other part of the conditional if it was `null`.

sam-github replied:
To be more clear, shouldn't it be `if(address != undefined && typeof address != 'string')` --- that is, any of the 3 values `undefined`, `null`, or String would be acceptable.

mscdex replied:
I'm not sure what you mean by 'acceptable,' but what I was saying is that

```js
address !== undefined && typeof address !== 'string'
```

and

```js
address != undefined && typeof address !== 'string'
```

will both evaluate to `true` for `address === null`.

boneskull replied:
I'm going to change this to:

```js
} else if (address !== undefined && typeof address !== 'string') {
```

mscdex commented on lib/dgram.js:
```diff
@@ -344,6 +346,13 @@ Socket.prototype.send = function(buffer,
+    callback = address;
+    address = undefined;
+  } else if (typeof address !== 'undefined' && typeof address !== 'string') {
+    throw new TypeError('"Address" argument must be a string or undefined');
```
Even though it goes against the common error message formatting used in the rest of node core, I think changing this to `'Address must be a string or undefined'` might be better since it is more consistent with the error message format when validating the `port`.

boneskull replied:
I'll take a closer look at the wording and punctuation.  IMO we should be shooting for consistency with core error messaging; not inconsistent error messaging which may be used elsewhere in this module.

sam-github commented on doc/api/dgram.md:
```diff
@@ -239,8 +239,9 @@ respect to [byte length][] and not the character position.
-DNS will be used to resolve the address of the host. If the `address` is not
-specified or is an empty string, `'127.0.0.1'` or `'::1'` will be used instead.
+DNS will be used to resolve the address of the host. Otherwise, if `address` is
+not specified or is an empty string, `'127.0.0.1'` (`udp4`) or `'::1'` (`udp6`)
```
What does "not specified" mean? I think in js we "provide" function arguments, not "specify" them. That aside, I don't understand from the text the behaviour of:

- `send("msg", 7)` // "not specified", I think
- `send("msg", 7, ()=>{}) // also not specified, I think
- `send("msg", 7, undefined) // specified as undefined, and handled by code below, I think
- `send("msg", 7, undefined, ()-.{}) // ditto
- `send("msg", 7, "") // same as if address is undefined
- `send("msg", 7, null) // not same as if address is `undefined`? This is weird and inconsistent, if true

I would state that if address is not provided, or is null, undefined, or an empty string, it will use the protocol specific localhost address - and implement and test this condition if it is not already

boneskull replied:
@sam-github I think @mscdex addressed your comments on line 352.  Not sure why I'm seeing it twice here.

boneskull replied:
I can change the wording to "provide". 

For an `address` that is not provided, we defer to the accepted types in `dns.lookup()`.  Per the code there, `address` is allowed to be a *falsy* value, or a string.

So, I may have been a bit too stringent in my checks.  The allowed values of `address` are then:

1.  a nonempty string (which is considered "provided"; the rest are not)
2. `false`
3. `undefined`
4. `null`
5. `0`
6. `''`

If it's any other truthy value (except `Function`; see below), it's invalid, and `send()` should throw a `TypeError`.

Furthermore, if `address` *is* a `Function`, then it's the `callback`, and the `address` will be reset to a falsy value (I think it was `undefined`, but it doesn't really matter).   

"Falsy" and "truthy" seem to be unique to JS and there's precedence in the Node.js docs for their usage.  Explicitly enumerating all "falsy" values is cumbersome, anyway.

I propose the following changes, then:

1.  Avoid ambiguity by stating "If `address` is a non-empty string, use it to resolve the host.  If `address` is falsy, `127.0.0.1` (etc) will be used instead.`
2.  Update my type check(s) to allow all falsy values to pass through to `dns.lookup()`.
3.  Revisit tests to ensure the "falsy" values *don't* throw exceptions, and truthy non-strings *do* throw exceptions.

Sound fair?

boneskull replied:
@sam-github @mscdex ^

mscdex replied:
@boneskull I don't see the usefulness in permitting values like `false` and `0` and treating them as `undefined`, if that is what you are suggesting?

sam-github replied:
I think he is saying that is current behaviour, and not doing it would be semver-major. I would say if we are going to change the API, I would expect only allow `null` and `undefined` to mean "no value". Btw, from above, it looks like `42` would also be considered a host name? I.e., numbers are allowed? Do they get converted to strings, like `'42'` and passed to the resolver?

boneskull replied:
> @boneskull I don't see the usefulness in permitting values like false and 0 and treating them as undefined, if that is what you are suggesting?

This is the current behavior.  Values are passed through to `dns.lookup()`, which allows such things.  The documentation for `dgram` is therefore inaccurate.  And ambiguous, which prompted the change.

> I think he is saying that is current behaviour, and not doing it would be semver-major. I would say if we are going to change the API, I would expect only allow null and undefined to mean "no value"

This is correct.  I'm not proposing making a semver-major change here.  Whether or not allowing `false` is "useful", I don't know, but I'd argue that's not a big enough deal to warrant a semver-major change.

> Btw, from above, it looks like 42 would also be considered a host name? I.e., numbers are allowed? Do they get converted to strings, like '42' and passed to the resolver?

No.  The only valid `number` is `0`, which is falsy.  Then:

> If it's any other truthy value (except Function; see below), it's invalid, and send() should throw a TypeError.

sam-github commented:
Btw, labelled as minor because of

> Upon suggestion by @mscdex, I've modified the changes to make address optional in all cases.

but perhaps it should be major because of 

> more helpful error messaging in case when address argument is present but invalid; exceptions no longer fall through to dns module.

And of course if it starts rejecting previously valid input, it must be major, but that's still under discussion, I think.

boneskull commented:
> more helpful error messaging in case when address argument is present but invalid; exceptions no longer fall through to dns module.
> And of course if it starts rejecting previously valid input, it must be major, but that's still under discussion, I think.

It doesn't.  Either would throw an error; the errors thrown in this PR are more helpful/obvious.

sam-github commented:
@boneskull Changes in error message text are considered semver-major, because they can break apps that match against `err.message`. Its an API breakage, but not controversial, better error messages are a good thing. So, sounds like this is semver-major?

sam-github commented:
@boneskull FYI, I would have done this as one commit introducing tests, because the test coverage wasn't there over the invalid addresses, asserting the current behaviour, then a follow-on commit that improves the error messages. That would as a side-effect also make whether it is an API change or not crystal clear - because changes to test behaviour are API changes. Its probably too much work to rebase it like that, but a 2-phase approach for code that has no test coverage ATM is easier to review.

boneskull commented:
> @boneskull Changes in error message text are considered semver-major, because they can break apps that match against err.message. Its an API breakage, but not controversial, better error messages are a good thing. So, sounds like this is semver-major?

That's an unfortunate--but likely necessary--policy.  

I want to make it clear that I'm not interested in expanding this PR to further restrict allowed types or values of the `address` parameter (e.g., disallowing `false`).  If we go down that road, the same logic suggests we should also change `dns.lookup()`.  Basically, this would be much more effort than I had intended to give this PR.

boneskull commented:
> Its probably too much work to rebase it like that, but a 2-phase approach for code that has no test coverage ATM is easier to review.

This is actually pretty straightforward to do.

sam-github commented:
I understand. This PR doesn't change the behaviour, other than making the error messages much more helpful.

The tests and docs do reveal that the current behaviour is not ideal, and now that you have done this work, its easier to discuss how we wish the API would work, and change it in the future.

boneskull commented:
> I understand. This PR doesn't change the behaviour, other than making the error messages much more helpful.

That's not entirely true either.  This PR *also* allows `callback()` without `address`.  It changes the API of `Socket.prototype.send()` in a non-breaking manner--i.e., it adds w/o taking away.

boneskull commented:
> The tests and docs do reveal that the current behaviour is not ideal, and now that you have done this work, its easier to discuss how we wish the API would work, and change it in the future.

Yes.  This is silly, but would work:

```js
socket.send(buf, port, '', () => {
  console.log('address was 127.0.0.1');
});
```

What's sillier is that currently you'd have to write something like the above code if you wanted a callback and didn't care to specify the address.  😝

boneskull commented:
To recap, my plan is:

- [ ] Change some verbiage in docs around "specify"
- [ ] Split PR into two commits:
  - [ ] tests against currently uncovered & undocumented behavior
  - [ ] implementation of and more tests around new behavior

boneskull commented:
@mscdex @sam-github Please take another gander.

boneskull commented:
I also realized this squashes a particularly gross bug--if `send()` was called with an invalid address type, but the socket was not yet fully bound, the send would be *enqueued*.  Once binding was complete, the send would run, and the exception from `dns.lookup()` would be thrown asynchronously.

You can see this problem illustrated in the tests of my first commit.  I was not able to assert the exceptions were thrown unless I did this (or caught them at the `process` level).  I removed the need to bind up-front in the second commit, since my changes fix the bug.

sam-github commented on doc/api/dgram.md:
```diff
@@ -239,8 +239,9 @@ respect to [byte length][] and not the character position.
-specified or is an empty string, `'127.0.0.1'` or `'::1'` will be used instead.
+DNS will be used to resolve the address of the host.  If `address` is not 
+provided or otherwise falsy, `'127.0.0.1'` (`udp4`) or `'::1'` (`udp6`) will 
+be used by default.
```
> .... depending on whether socket is `'udp4'` or `'udp6'`.

^--- I think its worth stating explicitly how it the `or` is determined.

boneskull replied:
@sam-github this is done

sam-github commented on test/parallel/test-dgram-send-callback-buffer-length-empty-address.js:
```diff
@@ -0,0 +1,19 @@
+const dgram = require('dgram');
+const client = dgram.createSocket('udp4');
+
+const buf = Buffer.allocUnsafe(256);
```
The test doesn't rely on sending unspecified data, does it? Can you just use `const buf = Buffer.alloc(256, 'x');` as in the other test?

boneskull replied:
Sure; was just copypasta from yet another test...

boneskull replied:
@sam-github this is done

boneskull commented:
rebased onto `master` + linted

boneskull commented:
I have a question; this doesn't seem to be covered in `CONTRIBUTING.md`.

Since this is tagged `semver-major`, does it need to wait until v8.0.0 is about to be published to be merged?  Or should I be retargeting this to a different branch?  Do I need to do anything else?  

Thanks.

jasnell commented:
It can be merged into master. It just would not go into a release until v8.0.0

boneskull commented:
@jasnell Is there something holding up the merge?

gibfahn commented:
@mscdex LGTY?

mscdex commented on lib/dgram.js:
```diff
@@ -344,6 +346,14 @@ Socket.prototype.send = function(buffer,
+    address = undefined;
+  } else if (address && typeof address !== 'string') {
+    throw new TypeError('Invalid arguments: address must be a nonempty ' +
+      'string or falsy');
```
This should align with the string above, like:

```js
throw new TypeError('Invalid arguments: address must be a nonempty ' +
                    'string or falsy');
```

cjihrig replied:
@boneskull ping ^

boneskull replied:
there ought to be a lint check here

mscdex replied:
/cc @Trott ?

Trott replied:
(Feel free to ignore this. It isn't really about this PR. It's answering an @-mention, that's all.)

@mscdex There is a lint rule checks that arguments are aligned. That's one argument broken across multiple lines so the rule doesn't kick in. If we enforced alignment in that scenario, it would have all sorts of undesirable consequences. (Consider an inline callback. The body of the function would need to be aligned with other arguments before the callback.)

All of this would be more manageable if we either had more lax rules around line length (something I would support--I think breaking a string apart like that does more harm to readability than letting it run off the right side of your editor) or if we didn't enforce alignment and instead had a constant indent (also something I would support), but those ships have probably sailed.

cjihrig commented on test/parallel/test-dgram-send-address-types.js:
```diff
@@ -0,0 +1,48 @@
+// invalid address: object
+assert.throws(() => {
+  client.send(buf, port, []);
+}, TypeError);
```
Instead of just checking for the `TypeError` constructor, we've been moving toward a check like this:

`/^TypeError: Invalid arguments: address must be a nonempty string or falsy$/`

You can put it in a variable and reuse it across these assertions.

boneskull commented:
I'm hoping to offer some constructive criticism. 

This is my first PR into Node. There were a couple rounds of changes soon after I sent it. These were significant changes to the PR. It appeared I had approval soon after this. 

Some weeks later, more changes were requested. 

This last round of changes seem like nitpicks (for lack of a better word), in my opinion. You may feel differently, but the relative impact on the PR as a whole is minimal. 

Many potential contributors would become extremely frustrated after having the PR sit for so long.

Requesting change upon change is something I try to avoid, as a maintainer. It's discouraging. I'll typically merge a PR at "good enough" instead of "perfect" (I think Paul Fenwick gave a talk on this...). 

Anyway, I'll make the changes, because I want to see this land. But please consider my feedback. 🙂

Trott commented:
> I'm hoping to offer some constructive criticism.

Thanks for the thoughtful feedback.

This project has around 90 people with commit access. With that many people, it can be a challenge to have PR interactions be consistent, positive, and timely. Getting constructive feedback when people have less-than-stellar experiences is really useful and I appreciate you taking the time to do so.

Trott commented:
> @jasnell Is there something holding up the merge?

semver-major changes require two approvals from CTC members. This so far has one (from @jasnell). If @cjihrig's nit is addressed and they change their review to approved, that would be enough for it to land.

Trott commented:
(Oh and this also will need a run through the continuous integration server to get tested across platforms. In fact, even though there are changes coming, I'll do that now just to surface any surprises like "OMG this doesn't work at all on Windows!!!!" or something like that.)

Preliminary CI: https://ci.nodejs.org/job/node-test-pull-request/6299/

boneskull commented:
I'm seeing a failure on ARM: 

![image](https://cloud.githubusercontent.com/assets/924465/22762045/3065d218-ee12-11e6-902c-50292262b048.png)

but no corresponding failure [on Jenkins](https://ci.nodejs.org/job/node-test-commit-arm/7635/).

Trott commented:
@boneskull Sorry about that. The arm failure in the widget is a known bug. CI is green.

boneskull commented:
I've rebased onto `master`, added the whitespace formatting requested by @mscdex and used the testing strategy as suggested by @cjihrig.  See 5940f11.

Trott commented:
CI for the new changes: https://ci.nodejs.org/job/node-test-pull-request/6300/

boneskull commented:
Hm, linty rebase.

boneskull commented:
Trying again with 61fdbd8.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6305/

boneskull commented:
arm failure again is unrelated

@cjihrig would appreciate a 👍  

😄

boneskull commented:
conflict resolved

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6346/

Trott commented:
I'm trying to run CI against this PR and it is not going well. I'm pretty sure the problem is with CI and not this PR, though. Maybe someone else seeing this in another hour or two can try again?

boneskull commented:
@Trott Anything I can do to help?

Trott commented:
@boneskull Yes! Can you remove the merge commit? Rumor has it that is the source of the issue.

boneskull commented:
@Trott Hm, that's what I get for using GitHub's new conflict resolver tool, I guess. :smile:

boneskull commented:
@Trott rebased

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6397/

gibfahn commented:
>This last round of changes seem like nitpicks (for lack of a better word), in my opinion. You may feel differently, but the relative impact on the PR as a whole is minimal.

@boneskull This is fair, but not being strict about things like https://github.com/nodejs/node/pull/10473#discussion_r100142756 have caused breakage in the past.

I think you're right though, what we need is to have all this stuff covered by linter rules, so that everything can all be fixed at once.

<br/>

Speaking of which, regarding this (by @cjihrig):

>Instead of just checking for the `TypeError` constructor, we've been moving toward a check like this:
>
>`/^TypeError: Invalid arguments: address must be a nonempty string or falsy$/`
>
>You can put it in a variable and reuse it across these assertions.

This seems like something we should lint for in `test/`, i.e. we should require the second parameter in `assert.throws()` to be a regex starting with `^` and ending with `$`.  (@not-an-aardvark @Trott @silverwind @targos)

Trott commented:
>  i.e. we should require the second parameter in assert.throws() to be a regex starting with ^ and ending with $. 

We should definitely allow validation functions as an alternative too. It's just constructors that can be problematic.

There's more to talk about regarding this (specifically if we might be moving towards a situation where error message changes are not breaking changes anymore thanks to the error identifiers that @jasnell has introduced, and if so then whether that means fully-matching regexps will become anti-patterns as the second argument). But this is not the place for that discussion. (Personally, I think we're probably a long way off from being able to treat message changes as non-breaking changes, but I'm often wrong.)

not-an-aardvark commented:
One potential concern about linting for error constructors is that the rule might report an error for this code:

```js
const ERROR_PATTERN = /^Error: Something bad happened$/;

// as far as the rule is concerned, ERROR_PATTERN could be a constructor
assert.throws(foo, ERROR_PATTERN);
assert.throws(bar, ERROR_PATTERN);
```

It would be relatively simple to check for constructors that are known beforehand, e.g. `TypeError`.

Trott commented:
/cc @nodejs/ctc 

This semver-major has sufficient approvals to land, but it's been a sufficiently long time coming that it might be off the radar for some folks. PTAL if you get a chance.

thefourtheye commented on test/parallel/test-dgram-send-callback-multi-buffer-empty-address.js:
```diff
@@ -0,0 +1,27 @@
+
+const client = dgram.createSocket('udp4');
+
+const messageSent = common.mustCall(function messageSent(err, bytes) {
```
`assert.ifError(err)`

thefourtheye commented on test/parallel/test-dgram-send-callback-multi-buffer-empty-address.js:
```diff
@@ -0,0 +1,27 @@
+  client.send([buf1, buf2], port, messageSent);
+});
+
+client.on('message', common.mustCall(function onMessage(buf, info) {
```
Unused variable `info`.

thefourtheye commented on test/parallel/test-dgram-send-callback-buffer-length-empty-address.js:
```diff
@@ -0,0 +1,19 @@
+const offset = 20;
+const len = buf.length - offset;
+
+const onMessage = common.mustCall(function messageSent(err, bytes) {
```
`err` has to be asserted first.

boneskull replied:
FWIW, these two files are copypasta from `test-dgram-send-callback-buffer-length` which has the same issues.

Trott commented:
OMG the `test/arm` results display is fixed! :-D

boneskull commented:
@thefourtheye Changes applied

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/6415/

boneskull commented:
ping anyone

cjihrig commented on test/parallel/test-dgram-send-address-types.js:
```diff
@@ -0,0 +1,51 @@
+  client.send(buf, port, true);
+}, expectedError);
+
+client.unref();
```
I just noticed this. Is there any chance that the socket would be unref'ed and the process exit before all six messages are received. I am able to make the test fail artificially by adding a short timeout around the last successful send. It doesn't seem to be a problem on the CI, but wouldn't want it to be a source of future flakiness either.

Trott commented:
Landed in 32679c73c11c618d027095d96b00f2a77b243d4d.
Thanks for the contribution! 🎉
